backend/pikpak/pikpak.go

```diff
@@ -0,0 +1,1749 @@
+		Endpoint:    &endpoint,
+	}
+	sess, err := session.NewSession(cfg)
+	if err != nil {
+		return
+	}
+
+	uploader := s3manager.NewUploader(sess)
+	// Upload input parameters
+	uParams := &s3manager.UploadInput{
+		Bucket: &p.Bucket,
+		Key:    &p.Key,
+		Body:   in,
+	}
+	// Perform upload with options different than the those in the Uploader.
+	_, err = uploader.UploadWithContext(ctx, uParams, func(u *s3manager.Uploader) {
+		// TODO can be user-configurable
+		u.PartSize = 10 * 1024 * 1024 // 10MB part size
+	})
+	return
+}
+
+func (f *Fs) upload(ctx context.Context, in io.Reader, leaf, dirID string, size int64, options ...fs.OpenOption) (*api.File, error) {
+	// Calculate sha1sum; grabbed from package jottacloud
+	//
+	// unwrap the accounting from the input, we use wrap to put it
+	// back on after the buffering
+	var wrap accounting.WrapFn
+	in, wrap = accounting.UnWrap(in)
+	var cleanup func()
+	sha1Str, in, cleanup, err := readSHA1(in, size, int64(f.opt.HashMemoryThreshold))
+	defer cleanup()
+	if err != nil {
+		return nil, fmt.Errorf("failed to calculate SHA1: %w", err)
+	}
+	// Wrap the accounting back onto the stream
+	in = wrap(in)
+
+	// determine upload type
+	uploadType := api.UploadTypeResumable
+	if size >= 0 && size < int64(5*fs.Mebi) {
+		uploadType = api.UploadTypeForm
+	}
+
+	// request upload ticket to API
+	req := api.RequestNewFile{
+		Kind:       api.KindOfFile,
+		Name:       f.opt.Enc.FromStandardName(leaf),
+		ParentID:   parentIDForRequest(dirID),
+		FolderType: "NORMAL",
+		Size:       size,
+		Hash:       strings.ToUpper(sha1Str),
+		UploadType: uploadType,
+	}
+	if uploadType == api.UploadTypeResumable {
+		req.Resumable = &map[string]string{"provider": "PROVIDER_ALIYUN"}
+	}
+	newfile, err := f.requestNewFile(ctx, &req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a new file: %w", err)
+	}
+	if newfile.File == nil {
+		return nil, fmt.Errorf("invalid response: %+v", newfile)
+	} else if newfile.File.Phase == api.PhaseTypeComplete {
+		// early return; in case of zero-byte objects
+		return newfile.File, nil
+	}
+
+	if uploadType == api.UploadTypeForm && newfile.Form != nil {
+		err = f.uploadByForm(ctx, in, req.Name, size, newfile.Form, options...)
+	} else if uploadType == api.UploadTypeResumable && newfile.Resumable != nil {
+		err = f.uploadByResumable(ctx, in, newfile.Resumable, options...)
+	} else {
+		return nil, fmt.Errorf("unable to proceed upload: %+v", newfile)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to upload: %w", err)
+	}
+	// refresh uploaded file info
+	return f.getFile(ctx, newfile.File.ID)
+}
+
+// PutUnchecked the object into the container
+//
+// This will produce an error if the object already exists.
+//
+// Copy the reader in to the new object which is returned.
+//
+// The new object may have been created if an error is returned
+func (f *Fs) PutUnchecked(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.Object, error) {
```

Comment:
If duplicates aren't possible, you don't need this method at all.

Reply:
Check the drive backend for a backend which can have duplicate file names.

If you call PutUnchecked on drive twice you will get two files with the same name, whereas if you call Put twice you will get one file and the second file updates the first file. This is what users expect!

Are duplicates possible?

Reply:
If duplicates aren't possible then you can remove this method (or merge it into Put/Putstream like it is in the other backends).

You can also remove MergeDirs